File: lc_java/d/Word Break II.java
/**
 * Given a string s and a dictionary of words dict, add spaces in s to construct a sentence where each word is a valid dictionary word.
 * Return all such possible sentences.
 * For example, given
 * s = "catsanddog",
 * dict = ["cat", "cats", "and", "sand", "dog"].
 * A solution is ["cats and dog", "cat sand dog"].
 */

public class Solution {
  public List<String> wordBreak(String s, Set<String> wordDict) {
    List<Set<Integer>> backtrace = new ArrayList<>();
    for (int i = 0; i < s.length(); i++) backtrace.add(new HashSet<>());
    for (int i = 0; i < s.length(); i++) {
      if (i == 0 || backtrace.get(i - 1).size() > 0) { // This condition is an optimization
        for (int j = i + 1; j <= s.length(); j++) {
          String w = s.substring(i, j);
          if (wordDict.contains(w)) {
            backtrace.get(j - 1).add(i);
          }
        }
      }
    }
    List<String> ret = new ArrayList<>();
    dfs(ret, s, backtrace, backtrace.size() - 1, "");
    return ret;
  }

  void dfs(List<String> ret, String s, List<Set<Integer>> bt, int i, String path) {
    if (i == -1) ret.add(path.substring(1));
    else {
      for (int beg : bt.get(i)) {
        dfs(ret, s, bt, beg - 1, " " + s.substring(beg, i + 1) + path);
      }
    }
  }
}

File: lc_plusplus/Word Break II.cc
// Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, add spaces in s to construct a sentence where each word is a valid dictionary word. You may assume the dictionary does not contain duplicate words.
//
// Return all such possible sentences.
//
// For example, given
// s = "catsanddog",
// dict = ["cat", "cats", "and", "sand", "dog"].
//
// A solution is ["cats and dog", "cat sand dog"].
//


class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> d{wordDict.begin(), wordDict.end()};
        vector<vector<int>> dp(s.size());
        for (size_t i = 0; i < s.size(); ++i) {
            if (i == 0 || !dp[i - 1].empty()) {
                for (size_t j = i; j < s.size(); ++j) {
                    if (d.count(s.substr(i, j - i + 1)) == 1) dp[j].push_back(i);
                }
            }
        }
        vector<string> res, words;
        dfs(res, words, dp.size() - 1, dp, s);
        return res;
    }

    void dfs(vector<string>& res, vector<string>& words, int idx, vector<vector<int>>& dp, string& str) {
        if (idx == -1) {
            string s;
            for (auto it = words.rbegin(); it != words.rend(); ++it) s += *it + " ";
            s.pop_back();
            res.push_back(s);
            return;
        }
        for (auto i : dp[idx]) {
            words.push_back(str.substr(i, idx - i + 1));
            dfs(res, words, i - 1, dp, str);
            words.pop_back();
        }
    }
};
